src/Type/Php/ArrayMergeFunctionDynamicReturnTypeExtension.php

```diff
@@ -26,16 +28,54 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
 
 	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
 	{
-		if (!isset($functionCall->getArgs()[0])) {
+		$args = $functionCall->getArgs();
+
+		if (!isset($args[0])) {
 			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
 		}
 
+		$argTypes = [];
+		$allConstant = true;
+		foreach ($args as $arg) {
+			$argType = $scope->getType($arg->value);
+			$argTypes[] = $argType;
+
+			if ($arg->unpack || !$argType instanceof ConstantArrayType) {
+				$allConstant = false;
+				break;
+			}
+		}
+
+		if ($allConstant) {
+			$newArrayBuilder = ConstantArrayTypeBuilder::createEmpty();
+			foreach ($argTypes as $argType) {
+				if (!$argType instanceof ConstantArrayType) {
+					throw new ShouldNotHappenException();
```

Comment:
You could push only ConstantArrayType into `$argTypes` and this condition would not be necessary :)

Reply:
but then I couldn't re-use the $artTypes for the 2nd-foreach loop..?